src/Aspire.Hosting/Health/ResourceHealthCheckService.cs

```diff
@@ -13,21 +13,57 @@ namespace Aspire.Hosting.Health;
 
 internal class ResourceHealthCheckService(ILogger<ResourceHealthCheckService> logger, ResourceNotificationService resourceNotificationService, HealthCheckService healthCheckService, IServiceProvider services, IDistributedApplicationEventing eventing, TimeProvider timeProvider) : BackgroundService
 {
-    private readonly Dictionary<string, ResourceEvent> _latestEvents = new();
-    private readonly Dictionary<string, ResourceMonitorState> _resourcesStartedMonitoring = new();
+    private readonly Dictionary<string, ResourceMonitorState> _resourceMonitoringStates = new();
 
-    internal class ResourceMonitorState(CancellationToken serviceStoppingToken)
+    internal class ResourceMonitorState(ResourceEvent initialEvent, CancellationToken serviceStoppingToken)
     {
         private readonly CancellationTokenSource _cts = CancellationTokenSource.CreateLinkedTokenSource(serviceStoppingToken);
+        private TaskCompletionSource? _delayInterruptTcs;
+        private CancellationTokenSource? _delayCts;
 
         public CancellationToken CancellationToken => _cts.Token;
 
+        public ResourceEvent LatestEvent { get; private set; } = initialEvent;
+
         public void StopResourceMonitor()
         {
             _cts.Cancel();
         }
+
+        public void SetLatestEvent(ResourceEvent resourceEvent)
+        {
+            LatestEvent = resourceEvent;
+            _delayInterruptTcs?.TrySetResult();
+        }
+
+        internal async Task<bool> DelayAsync(TimeSpan delay, CancellationToken cancellationToken)
+        {
+            if (_delayCts == null || !_delayCts.TryReset())
+            {
+                _delayCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            }
+            _delayInterruptTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            var completedTask = await Task.WhenAny(Task.Delay(delay, _delayCts.Token), _delayInterruptTcs.Task).ConfigureAwait(false);
```

Comment:
This is old school. We have `WaitAsync` now. It throws an exception but it looks cleaner 😄